### ISSUE TITLE
`onCompleted` gets called after loading state has been changed.. 

### ISSUE THREAD

**Issue by novaknole** [apollographql/apollo-client#8350]
Hi, I kind of find it hard to follow the template and I believe the way I describe the issue should help you better understand the problem.

This is what my function looks like 

```

export function useDaoQuery(daoName: string) {
  const [daoData, setDaoData] = useState<any>(null);
  const { data, loading, error } = useQuery(DAO_BY_NAME, {
    variables: { name: daoName },
    onCompleted: () => {
      const daoData = transformDaoDetails(data); // just some quick transformation function
      setDaoData(daoData);
    },
  });

  return {
    data: daoData,
    loading,
    error,
  };
}
```

As shown, I return `data`, `loading` and `error`.. 

The problem with the above is that:

1st render - `loading` = true, `daoData` = null
2nd render - `loading` = false, `daoData` = null
3th render - `loading` = false, `daoData` = `actual object`.

The problem is the 2nd render. This happens because of `onCompleted`.   What I need is when `loading` is false, `daoData` should also be the ready data, but instead, as soon as loading gets false,  that's when it starts `onCompleted` and I end up with 2nd render above and this is truly a problem because I can't really figure out when to show `not found` record and when to show `loading`..

The only solution I can think of is bringing my `isLoading` variable and setting it to `false` in the `onCompleted` hook when everything is done.  Is this the right choice? but wouldn't I need to do the same with `onError` callback too ? which seems really bad to me. 

Thanks a lot.

**Comment by dylanwulf** [comment]
What about something like this?
```js
export function useDaoQuery(daoName: string) {
  const { data, loading, error } = useQuery(DAO_BY_NAME, {
    variables: { name: daoName },
  });
  const daoData = useMemo(() => transformDaoDetails(data), [data]);

  return {
    data: daoData,
    loading,
    error,
  };
}
`

**Comment by novaknole** [comment]
@dylanwulf 

Yes, In the end, This is what I am going to do.

But, I also wanted to make sure this is not a bug, but how it should work with `onCompleted`...

**Comment by dylanwulf** [comment]
@novaknole Take a look at #8137. That issue deals with `useMutation` instead of `useQuery` but otherwise sounds a lot like what you're describing.

**Comment by ohdoski** [comment]
Have you checked `NetworkStatus`?

https://www.apollographql.com/docs/react/data/queries/#inspecting-loading-states

**Comment by amerikan** [comment]
It seems like `loading` is changed to `true` and then `onCompleted` gets called. I had something setup along these lines, though not sure if it should work or not (it doesn't I get undefined) during the first render. Unless it's not meant to be used in this manner.

```js
const [animals, setAnimals] = useState([])

  const { loading, error, data } = useQuery(
    GET_ANIMALS,
    {
      onCompleted: (animalData) => {
        setAnimals(animalData); // state
      }
    }
  );

  if (loading) {
    return "Loading..."
  }

  console.log(animals); // undefined <- should have the animals by this point?
  console.log(data); // ["pig", "dog", "cat"]
```

**Comment by CrossTheDev** [comment]
I usually solve this by implementing custom loading. I set it to false inside the onCompleted.

```javascript
const [loading, setLoading] = useState(true)
useSomeQuery({
    onCompleted: (data) => {
      ...
      setLoading(false)
    },
  })

**Comment by alessbell** [comment]
Hi @novaknole 👋  This issue is a few years old now, and we recently fixed a bug related to `onCompleted` that spurred discussion of this callback function and when it should be utilized.

Without a reproduction I can't say whether this was originally a bug or not, but I wanted to add that for a few reasons you can (and I'd go farther by saying _should_) avoid `onCompleted` altogether here.

I say this because we fixed a regression that means that `onCompleted` is now no longer called for cache writes to the selection set being watched without an associated network request—since there's nothing to "complete"—but `data` can still update and cause a re-render.

By doing this computation inline in render, you also avoid waiting a render for `onCompleted` to be called before doing your data transformation, calling `setState` and then waiting for a subsequent re-render.

Here's what you can do instead:

```jsx
export function useDaoQuery(daoName: string) {
  const { data, loading, error } = useQuery(DAO_BY_NAME, {
    variables: { name: daoName }
  });

  const daoData = useMemo(
    () => data ? transformDaoDetails(data) : null,
    [data, transformDaoDetails] // don't forget to cache transformDaoDetails via useCallback too if needed
  );

  return {
    data: daoData,
    loading,
    error,
  };
}
```

Apollo's cache already ensures that `data` is referentially stable across renders, so this works nicely. Feel free to ask any questions over in our [Discord](https://discord.gg/graphos), but I'll go ahead and close this issue out :) Thanks!
